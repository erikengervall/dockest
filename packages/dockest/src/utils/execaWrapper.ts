import execa, { SyncOptions } from 'execa' // eslint-disable-line import/default
import { trim } from './trim'
import { Logger } from '../Logger'
import { Runner } from '../@types'

interface Opts {
  runner?: Runner
  logPrefix?: string
  logStdout?: boolean
  execaOpts?: SyncOptions<string>
}

export const execaWrapper = async (
  command: string,
  { runner, logPrefix = '[Shell]', logStdout = false, execaOpts = {} }: Opts = {},
) => {
  const trimmedCommand = trim(command)
  const logger = runner ? runner.logger : Logger

  logger.debug(`${logPrefix} <${trimmedCommand}>`)

  const result = execa.commandSync(trimmedCommand, {
    shell: true,
    ...execaOpts,
  })

  logStdout && logger.debug(`${logPrefix} Success (${result.stdout})`, { success: true })

  return result
}
