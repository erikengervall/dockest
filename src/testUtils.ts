import { createMockProxy } from 'jest-mock-proxy'
import execa from 'execa' // eslint-disable-line import/default
import { ZooKeeperRunner, KafkaRunner, PostgresRunner, RedisRunner, GeneralPurposeRunner } from './runners'
import { Runner } from './runners/@types'
import Logger from './Logger'
import { DEFAULT_USER_CONFIG, INTERNAL_CONFIG } from './constants'

export const mockedExecaStdout = 'getContainerId ðŸŒ®'
export const runnerCommand = 'runRunnerCommands ðŸŒ®'

/**
 * FIXME: Would love for this to work ðŸ¦Š
 */
// jest.mock('execa', () => jest.fn(() => ({ stdout: mockedExecaStdout })))
// jest.mock('./Logger')

const { values } = Object

const createDockestConfig = (opts: { runners?: Runner[] }) => {
  const { runners = [] } = opts

  for (const runner of runners) {
    runner.logger = createMockProxy()
  }

  return {
    runners,
    opts: DEFAULT_USER_CONFIG,
    jest: {},
    $: INTERNAL_CONFIG,
  }
}

export default ({ withRunnerCommands = false }) => {
  const withCmds = withRunnerCommands ? { commands: [runnerCommand] } : {}

  const zooKeeperRunner = new ZooKeeperRunner({ service: 'zookeepeer', image: 'some/image:123', ...withCmds })
  const kafkaRunner = new KafkaRunner({
    service: 'kafka',
    image: 'some/image:123',
    dependsOn: [zooKeeperRunner],
    ...withCmds,
  })
  const postgresRunner = new PostgresRunner({
    service: 'postgres',
    image: 'some/image:123',
    database: '_',
    username: '_',
    password: '_',
    ...withCmds,
  })
  const redisRunner = new RedisRunner({ service: 'redis', image: 'some/image:123', ...withCmds })
  const generalPurposeRunner = new GeneralPurposeRunner({ service: 'general', image: 'some/image:123', ...withCmds })
  const initializedRunners = {
    kafkaRunner,
    postgresRunner,
    redisRunner,
    zooKeeperRunner,
    generalPurposeRunner,
  }

  beforeEach(() => values(initializedRunners).forEach(runner => (runner.logger = createMockProxy())))

  return {
    initializedRunners,
    runners: {
      ZooKeeperRunner,
      KafkaRunner,
      PostgresRunner,
      RedisRunner,
      GeneralPurposeRunner,
    },
    execa,
    Logger,
    createDockestConfig,
  }
}
