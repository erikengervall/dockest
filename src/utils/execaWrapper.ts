import execa from 'execa'
import Logger from '../Logger'
import { Runner } from '../runners/@types'
import trim from '../utils/trim'

const execaWrapper = async (command: string, runner?: Runner): Promise<string> => {
  const trimmedCommand = trim(command)
  const logger = runner ? runner.logger : Logger

  logger.debug(`Executing shell script: ${trimmedCommand}`)
  const { stdout } = await execa(trimmedCommand, { shell: true })
  logger.debug(`Successfully executed shell script ${trimmedCommand}`)

  return stdout
}

export default execaWrapper
