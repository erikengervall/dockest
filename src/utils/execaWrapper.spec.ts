import execaWrapper from './execaWrapper'
import testUtils, { mockedExecaStdout } from '../testUtils'

const {
  initializedRunners: { redisRunner },
  Logger,
  execa,
} = testUtils({})

jest.mock('execa', () => jest.fn(() => ({ stdout: mockedExecaStdout })))
jest.mock('../Logger')

describe('execaWrapper', () => {
  const command = 'run some CLI command :please:'

  describe('with runner', () => {
    it('should work', async () => {
      const result = await execaWrapper(command, redisRunner)

      expect(redisRunner.logger.debug).toMatchSnapshot()
      expect(execa).toMatchSnapshot()
      expect(result).toMatchSnapshot()
    })
  })

  describe('without runner', () => {
    it('should work', async () => {
      const result = await execaWrapper(command)

      expect(Logger.debug).toMatchSnapshot()
      expect(execa).toMatchSnapshot()
      expect(result).toMatchSnapshot()
    })
  })
})
