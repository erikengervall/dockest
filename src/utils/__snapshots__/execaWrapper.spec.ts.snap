// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execaWrapper with runner should work 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Executing shell script: run some CLI command :please:",
    ],
    Array [
      "Successfully executed shell script run some CLI command :please:",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`execaWrapper with runner should work 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "run some CLI command :please:",
      Object {
        "shell": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "stdout": "getContainerId ðŸŒ®",
      },
    },
  ],
}
`;

exports[`execaWrapper with runner should work 3`] = `"getContainerId ðŸŒ®"`;

exports[`execaWrapper without runner should work 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "Executing shell script: run some CLI command :please:",
    ],
    Array [
      "Successfully executed shell script run some CLI command :please:",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`execaWrapper without runner should work 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "run some CLI command :please:",
      Object {
        "shell": true,
      },
    ],
    Array [
      "run some CLI command :please:",
      Object {
        "shell": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "stdout": "getContainerId ðŸŒ®",
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "stdout": "getContainerId ðŸŒ®",
      },
    },
  ],
}
`;

exports[`execaWrapper without runner should work 3`] = `"getContainerId ðŸŒ®"`;
